devices/ble_hci/common-hal/_bleio/Connection.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Dan Halbert for Adafruit Industries
+ * Copyright (c) 2018 Artur Pacholec
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#ifndef MICROPY_INCLUDED_BLE_HCI_COMMON_HAL_CONNECTION_H
+#define MICROPY_INCLUDED_BLE_HCI_COMMON_HAL_CONNECTION_H
+
+#include <stdbool.h>
+
+#include "py/obj.h"
+#include "py/objlist.h"
+
+#include "common-hal/_bleio/__init__.h"
+#include "shared-module/_bleio/Address.h"
+#include "common-hal/_bleio/Service.h"
+
+typedef enum {
+    PAIR_NOT_PAIRED,
+    PAIR_WAITING,
+    PAIR_PAIRED,
+} pair_status_t;
+
+// We split the Connection object into two so that the internal mechanics can live outside of the
+// VM. If it were one object, then we'd risk user code seeing a connection object of theirs be
+// reused.
+typedef struct {
+    uint16_t conn_handle;
+    bool is_central;
+    // Remote services discovered when this peripheral is acting as a client.
+    mp_obj_list_t *remote_service_list;
+    // The advertising data and scan response buffers are held by us, not by the SD, so we must
+    // maintain them and not change it. If we need to change the contents during advertising,
+    // there are tricks to get the SD to notice (see DevZone - TBS).
+    bonding_keys_t bonding_keys;
```

Comment:
Is this reusing the struct from the SD?

Reply:
This is our own definition, but it used the SD structs internally. It's actually empty now:
```c
typedef struct {
    // ble_gap_enc_key_t own_enc;
    // ble_gap_enc_key_t peer_enc;
    // ble_gap_id_key_t peer_id;
} bonding_keys_t;
```

I could have commented a lot of this out completely, but I tried not commenting out everything, in order to allow some code to still stand uncommented. However, its doesn't really matter.